activerecord/test/cases/associations/belongs_to_associations_test.rb

```diff
@@ -1453,6 +1454,22 @@ def test_polymorphic_with_custom_name_touch_old_belongs_to_model
     assert_equal touch_time, car.reload.wheels_owned_at
   end
 
+  def test_polymorphic_stale_state_handles_nil_foreign_keys_correctly
```

Comment:
@byroot added a test. It's a bit indirect, but that's exactly situation in which we encounter the issue in our codebase.

Reply:
Test seems fine to me. I pushed a small refactoring though.